R/superPartition.R

```diff
@@ -0,0 +1,269 @@
+#'
+#' @seealso [partition()]
+#'
+#' @author Katelyn Queen, \email{kjqueen@@usc.edu}
+#'
+#' @export
+super_partition <- function(full_data,
+                            threshold = 0.5,
+                            cluster_size = 4000,
+                            partitioner = part_icc(),
+                            tolerance = .0001,
+                            niter = NULL,
+                            x = "reduced_var",
+                            .sep = "_") {
+
+  # ensure 0 < threshold < 1
+  if(0 > threshold | 1 < threshold) stop("Threshold must be between 0 and 1.")
+
+  # ensure data frame structure
+  full_data <- as.data.frame(full_data)
+
+  # if < cluster_size features, call regular partition
+  if(ncol(full_data) < cluster_size) {
+    message(paste0("Using `partition()` since there are < ", cluster_size, "features."))
+    return(partition(full_data, threshold = threshold))
+  }
+
+  # iteration counters
+  i <- j <- k <- l <- 0
+
+  # Function to go from cluster-specific column numbers to full data column numbers
+  ## full_data  - sample by probe expression data for all probes
+  ## small_data - sample by probe expression data for probes in specific cluster
+  ## modules   - indices from partition mapping key
+  ## return - indices from partition mapping key for full data
+  full_data_col_numbers <- function(full_data, small_data, modules) {
+
+    # vector of lists to return
+    return_mods <- numeric(length(modules))
+
+    for (i in seq_along(modules)) {
+      # create vector for full data indices
+      col_full <- numeric(length(unlist(modules[i])))
+
+      # get column names
+      col_small <- colnames(small_data)[unlist(modules[i])]
+
+      # get column indices in full data
+      for (j in seq_along(col_small)) {
+        col_full[j] <- grep(col_small[j], colnames(full_data))
+      }
+
+      # add list of indices to return vector
+      return_mods[i] <- list(col_full)
+    }
+    # to return
+    return_mods
+  }
+
+  # tracking variables
+  master_cluster <- data.frame(col_name = colnames(full_data), cluster = 1)
+  num_modules    <- 0
+
+  # use genie (fast agglomerative hierarchical clustering) to cluster data into cluster_size chunks
+  ## transpose data since genie clusters on rows
+  clust                  <- genieclust::genie(t(full_data),
+                                              k = ceiling(ncol(full_data)/cluster_size),
+                                              gini_threshold = 0.05)
+  master_cluster$cluster <- clust
+  clust_size             <- table(clust)
+
+  # Function to continually reduce clusters until largest cluster has less than 4,000 features
+  # cs - clust_size, numeric vector of cluster sizes
+  # fd - full_data, sample x feature
+  reduce_clust <- function(cs, fd) {
+    # bookkeeping
+    largest_clust <- max(master_cluster$cluster)
+    unique_vals   <- unique(master_cluster$clust)
```

Comment:
```suggestion
    unique_vals   <- unique(master_cluster$cluster)
```

you're using partial matching here but that is also a common source of bugs

Reply:
This was a typo! It's fixed in the latest commit